### DISCUSSION TITLE
Support the prompt caching for GCP Vertex AI

### DISCUSSION THREAD

**Discussion by wowmarcomei** [cline/cline#459] [Feature Requests]
Can it support  prompt caching in GCP Vertex AI models to achieve substantial cost savings? This would be of great assistance.

**Comment by wowmarcomei** [comment]
I found it has reserved the configuration in `src/shared/api.ts` for Vertex AI, but it isn't enabled yet. I tried to figure out whether Vertex AI API supports it or not from its official website, but I never found it. Maybe it doesn't support it yet?

```tsx
// Vertex AI
// https://cloud.google.com/vertex-ai/generative-ai/docs/partner-models/use-claude
export type VertexModelId = keyof typeof vertexModels
export const vertexDefaultModelId: VertexModelId = "claude-3-5-sonnet@20240620"
export const vertexModels = {
	"claude-3-5-sonnet@20240620": {
		maxTokens: 8192,
		contextWindow: 200_000,
		supportsImages: true,
		supportsPromptCache: false,
		inputPrice: 3.0,
		outputPrice: 15.0,
	},
	"claude-3-opus@20240229": {
		maxTokens: 4096,
		contextWindow: 200_000,
		supportsImages: true,
		supportsPromptCache: false,
		inputPrice: 15.0,
		outputPrice: 75.0,
	},
	"claude-3-haiku@20240307": {
		maxTokens: 4096,
		contextWindow: 200_000,
		supportsImages: true,
		supportsPromptCache: false,
		inputPrice: 0.25,
		outputPrice: 1.25,
	},
} as const satisfies Record<string, ModelInfo>

export const openAiModelInfoSaneDefaults: ModelInfo = {
	maxTokens: -1,
	contextWindow: 128_000,
	supportsImages: true,
	supportsPromptCache: false,
	inputPrice: 0,
	outputPrice: 0,
}
```

**Comment by josephomills** [comment]
@wowmarcomei The latest version of Clause Sonnet 3.5 (claude-3-5-sonnet@20241022) supports prompt caching and it's also enabled on Vertex AI (as mentioned on their [pricing page](https://cloud.google.com/vertex-ai/generative-ai/pricing#claude-models)). So this should be added so we save on costs and enjoy Cline. 
Great work BTW @cline!

**Comment by alliecatowo** [comment]
I'd love to see prompt caching supported with vertex ai! It's getting a bit pricey for me without it 😬

**Comment by dougnukem** [comment]
Appears to be available in the Vertex AI REST API:
https://cloud.google.com/vertex-ai/generative-ai/docs/partner-models/claude-prompt-caching
> Caching automatically occurs when subsequent requests contain the identical text, images, and cache_control parameter as the first request. All requests must also include the cache_control parameter in the same blocks.

**Comment by dougnukem** [comment]
PR that implements this:
https://github.com/cline/cline/pull/1885

**Comment by dougnukem** [comment]
@alliecatowo , @josephomills , @wowmarcomei While waiting you can use this pre-release version https://github.com/dougnukem/cline/releases/download/dougnukem-v3.4.8/cline-3.4.8.vsix

**Comment by wowmarcomei** [comment]
noted, thanks for your efforts and reply.
